### PR TITLE
Vfmadd

### PR DESCRIPTION
**Pull request by P1K**
Tentatively solving issue #274 (plus some minor change in `fflas_fscal.inl`)

This is mostly a duplication of `fflas_fscal.inl` into `fflas_faxpy.inl` and replacing muls by muladds.
Tested with and without vectorisation enabled on albaron (AVX2) and retourdest (AVX512)

### CODE REVIEW

File: fflas-ffpack/fflas/fflas_faxpy.inl [hunk 1]
```diff
@@ -27,18 +28,168 @@
...
+        }
+        else
+        {
+            if (N < FFLASFFPACK_COPY_REDUCE) // defined in fflas_freduce.inl
+            {
+                vectorised::axpyp(F,a,X,Y,N,incX,incY);
+            }
+            else
+            {
+                typename Field::Element_ptr Xc = fflas_new (F,N) ;
```
Comment (ClementPernet) [inline fflas-ffpack/fflas/fflas_faxpy.inl]: If only one of X or Y has an inc !=1 then this code copies both vectors anyway!

File: fflas-ffpack/fflas/fflas_faxpy.inl [hunk 2]
```diff
@@ -27,18 +28,168 @@
...
+
+    template<class Field>
+    inline typename std::enable_if<FFLAS::support_fast_mod<typename Field::Element>::value, void>::type
+    axpyp(const Field &F, const typename Field::Element a, typename Field::ConstElement_ptr X, typename Field::Element_ptr Y, const size_t n, const size_t incX, const size_t incY,
+              HelperMod<Field> &H)
+    {
+        typename Field::ConstElement_ptr Xi = X;
+        typename Field::Element_ptr Yi=Y;
+        for (; Xi < X+n*incX; Xi+=incX, Yi+=incY )
+            *Yi = reduce(a*(*Xi) + (*Yi), H);
```
Comment (ClementPernet) [inline fflas-ffpack/fflas/fflas_faxpy.inl]: Are we sure that the infix operators + and * are legit for this field?

File: fflas-ffpack/fflas/fflas_faxpy.inl [hunk 3]
```diff
@@ -27,18 +28,168 @@
...
+            {
+                vectorised::axpyp(F,a,X,Y,N,incX,incY);
+            }
+            else
+            {
+                typename Field::Element_ptr Xc = fflas_new (F,N) ;
+                typename Field::Element_ptr Yc = fflas_new (F,N) ;
+                fassign(F,N,X,incX,Xc,1);
+                fassign(F,N,Y,incY,Yc,1);
+                faxpy(F,N,a,Xc,1,Yc,1,FieldCategories::ModularTag());
```
Comment (P1K) [inline fflas-ffpack/fflas/fflas_faxpy.inl]: From JGD: why not a call to vectorised::axpyp directly?

File: fflas-ffpack/fflas/fflas_faxpy.inl (hunk 1 above)
Comment (P1K) [reply fflas-ffpack/fflas/fflas_faxpy.inl]: This looks indeed wasteful

File: fflas-ffpack/fflas/fflas_faxpy.inl (hunk 2 above)
Comment (P1K) [reply fflas-ffpack/fflas/fflas_faxpy.inl]: This should come from support_fast_mod (also implied by support_simd_mod)

Comment (ClementPernet) [reply fflas-ffpack/fflas/fflas_faxpy.inl]: Ok. At some point, one should consider cleaning up these traits. This is #304

Comment (ClementPernet) [reply fflas-ffpack/fflas/fflas_faxpy.inl]: This axpyp should also not be in the namespace vectorised as it does not use vectorization

File: fflas-ffpack/fflas/fflas_faxpy.inl (hunk 1 above)
Comment (P1K) [reply fflas-ffpack/fflas/fflas_faxpy.inl]: Should be fixed in latest commit

File: fflas-ffpack/fflas/fflas_faxpy.inl (hunk 3 above)
Comment (P1K) [reply fflas-ffpack/fflas/fflas_faxpy.inl]: Should be fixed in latest commit

**Comment by P1K** [comment]
Tests passed on `retourdest`but failed on `albaron`, so probably not ready

**Comment by P1K** [comment]
There is still at least one bug in this branch w.r.t. the current master, but I believe it's due to recent changes in cardinality bounds.
Essentially, the "fast" reduce on `int64_t` gives erroneous results when the to-be-reduced input overflowed as a product into a negative number (which is undefined behaviour, if I'm not wrong). (Then even if the overflow did not technically result in a loss of information, the computed result will be wrong due to difference between signed/unsigned division used in the reduce function.)
Example (from a real test):
```
int64_t a = 3315567202
int64_t b = 3606843909
int64_t c = a*b == -6488010706295678998 // unsigned would be 11958733367413872618 < 2**64-1
```
In short, I believe that the cardinality bound is incorrect and that no modulus should make such an overflow possible.

Note: Even if I encountered this problem while testing this branch, I see no reason why the same bug would not be currently present on master. Maybe it's simply better hidden.

@cyrilbouvier , @ClementPernet do you have an opinion on this?

**Comment by P1K** [comment]
The bug from my previous comment actually came from a not-up-to-date master (I am apparently too distracted and/or not skilful enough to actually merge master into branches).
Everything seems to be fine now except a fail on test-fger, but that one also happens on master (cf. #307)

File: fflas-ffpack/fflas/fflas_faxpy.inl [hunk 4]
```diff
@@ -172,14 +172,38 @@ namespace FFLAS { namespace details {
...
+                    Xc = const_cast<typename Field::Element_ptr>(X); // Oh the horror
+                }
+                if (incY != 1)
+                {
+                    Yc = fflas_new (F,N);
+                    fassign(F,N,Y,incY,Yc,1);
+                }
+                else
+                {
+                    Yc = const_cast<typename Field::Element_ptr>(Y);
```
Comment (ClementPernet) [inline fflas-ffpack/fflas/fflas_faxpy.inl]: No need for const_cast as Y is not const

File: fflas-ffpack/fflas/fflas_freduce.inl [hunk 5]
```diff
@@ -544,10 +544,49 @@ namespace FFLAS { namespace details {
...
+                typename Field::Element_ptr Ac;
+                typename Field::Element_ptr Bc;
+                if (incX != 1)
+                {
+                    Ac = fflas_new (F,m);
+                    fassign(F,m,A,incX,Ac,1);
+                }
+                else
+                {
+                    Ac = const_cast<typename Field::Element_ptr>(A); // Oh the horror
```
Comment (ClementPernet) [inline fflas-ffpack/fflas/fflas_freduce.inl]: Same here: A is not const, so why a  const_cast

File: fflas-ffpack/fflas/fflas_freduce.inl [hunk 6]
```diff
@@ -544,10 +544,49 @@ namespace FFLAS { namespace details {
...
+                // TODO also write a ``fast'' modp that handles an incY
+                typename Field::Element_ptr Xi = A ;
+                typename Field::ConstElement_ptr Yi = B ;
+                for (; Xi < A+m*incX; Xi+=incX, Yi += incY )
+                    F.reduce (*Xi , *Yi);
+            }
+            else
+            {
+                typename Field::Element_ptr Ac;
+                typename Field::Element_ptr Bc;
```
Comment (ClementPernet) [inline fflas-ffpack/fflas/fflas_freduce.inl:558]: Why do we need Bc to be Element_ptr and not ConstElement_ptr, which would then remove the need for const_cast

File: fflas-ffpack/fflas/fflas_faxpy.inl [hunk 7]
```diff
@@ -172,14 +172,38 @@ namespace FFLAS { namespace details {
...
             {
-                typename Field::Element_ptr Xc = fflas_new (F,N) ;
-                typename Field::Element_ptr Yc = fflas_new (F,N) ;
-                fassign(F,N,X,incX,Xc,1);
-                fassign(F,N,Y,incY,Yc,1);
-                faxpy(F,N,a,Xc,1,Yc,1,FieldCategories::ModularTag());
-                fassign(F,N,Yc,1,Y,incY);
-                fflas_delete(Xc);
-                fflas_delete(Yc);
+                typename Field::Element_ptr Xc;
```
Comment (ClementPernet) [inline fflas-ffpack/fflas/fflas_faxpy.inl:175]: Define Xc as ConstElement_ptr instead, which will remove the need for const_cast

Comment (P1K) [reply fflas-ffpack/fflas/fflas_faxpy.inl:175]: Cf. our offline discussion, the const_cast is needed for compatibility with the signature of `faxpy`. The proposed fix doesn't work (as is) since there is a branch where `Xc` is assigned

File: fflas-ffpack/fflas/fflas_faxpy.inl [hunk 8]
```diff
@@ -27,18 +28,192 @@
...
+        long st = long(Y) % simd::alignment;
+        if (st)
+        { // the array Y is not aligned (process few elements until (Y+i) is aligned)
+            for (size_t j = static_cast<size_t>(st) ; j < simd::alignment ; j+=sizeof(Element), i++)
+            {
+                Y[i] = reduce(a*X[i]+Y[i], H);
+            }
+        }
+        FFLASFFPACK_check((long(Y+i) % simd::alignment == 0));
+        if (long(Y+i)%simd::alignment==0)
```
[the remark this answers is not in the document]
Comment (P1K) [reply fflas-ffpack/fflas/fflas_faxpy.inl:77]: A: no, FFLASFFPACK_check is just a debug thingy so let it be

**Comment by P1K** [comment]
> The bug from my previous comment actually came from a not-up-to-date master (I am apparently too distracted and/or not skilful enough to actually merge master into branches).
> Everything seems to be fine now except a fail on test-fger, but that one also happens on master (cf. #307)

All tests now passing with master (2.5.0rc0) on `albaron` (AVX2) and `retourdest` (AVX512)

**Comment by ClementPernet** [comment]
Back to it. LGTM
